### ISSUE TITLE
Speeding up non-SSE version

### ISSUE THREAD

**Issue by Bulat-Ziganshin** [nkurz/countbench#1]
I just read you conversation in Yann blog, in particular this part:

> Number 2 "count2x64" is the non-SSE interleaved high/low byte approach, which at 3050 MB/s at 3.9GHz is running at 1.27 cycles/byte.
> 
> The fastest one, Number 20 "port7vec" at the bottom, is at 1.21 cycles/byte for 3200 MB/s. It's a ridiculous Haswell only AVX2 approach that converts the 16 input bytes at a time to a 16 x 16-bit vector, does a 2-bit shift left on the entire vector, then writes the vector to a fixed buffer. It's then reread as 16 separate scalar words, but since the shift has already been done the store can execute on Port 7. 

... and came with idea that may allow non-SSE version to become as fast as SSE one: make main frequency tables BYTE-wide and use jz/jc to increment supplementary table on overflows:

``` C
char freq0[256];
size_t main_freq[256];
for(...)
    if (++freq0[c] == 0)  main_freq[c]++;
```

Well, i don't know whether byte memory operations got a performance hit compared to operations with 32-bit entities

**Comment by nkurz** [comment]
Perhaps, although I'm not sure I follow your logic.  Can you explain more? 

> Well, i don't know whether byte memory operations got a performance hit 
> compared to operations with 32-bit entities

Some background:

8-bit, 16-bit, 32-bit, and 64-bit memory operations all require a a single instruction and a single µop and have the same latency.   Occasionally the length of the assembly instruction makes a difference, but rarely.     Vector loads add one cycle of latency, but are also one instruction and one µop.  

Certain "addressing forms" add 1 cycle of latency.   Usually, though, for an out-of-order processor like all modern Intel processors, latency isn't a problem, since the processor can find other instructions to execute.  

Addition operations have the same latency and same throughput for 8-bit, 16-bit, 32-bit, 64-bit, and all vector widths.    Using 8-bit math isn't any faster than using 64-bit.   Possibly this is what you were hoping?  

Add does pair with 'jc' so it's treated as a single µop internally.  See "macro fusion" in Agner's guides.  This should make it free when it doesn't fire.  On potential problem is that a mispredicted branch (when jc is true) costs you ~15 cycles penalty for the processor to recover.   

Very useful background reading if you are interested in these topics: 

http://www.agner.org/optimize/microarchitecture.pdf

http://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-optimization-manual.html

Anyway, feel free to explain more if you think you have a good idea.  Even better, perhaps you can test it and prove it!  I haven't thought about this code in a while, but I'm happy to look at it again.

**Comment by Bulat-Ziganshin** [comment]
yeah, i know microarchitecture at Agner/IntelManuals level, but have no experience of VTune profiling

If you have read my message in email, look at final edited variant above - i cited the point. You have used SSE/AVX to multiple each byte by 4 in order to simplify adressing mode. Instead, i propose to omit this multiplication and increment byte-wide frequency counters, touching wider counters only on overflow. It will lead to one unpredicted jump per 256 bytes processed, i.e. 14/256 cycles lost per each processed byte. Damn, it's 0.054 cycles, that's pretty close to 5% speed difference you have measured between AVX and non-SSE versions.

Well, on the second thought, it drops a few SIMD operations compared to your fastest code, so it still may be faster than best non-SSE code. Probably it can't buy anything for pre-Haswell cores, since they don't have this port7-problem.

**Comment by nkurz** [comment]
I looked at your actual message before responding, but I wasn't able to determine exactly where you meant to apply your changes.  

Although  maybe when you say "may allow non-SSE version to become as fast as SSE" you mean "allow the SSE version to become as fast the the AVX"?  

It's probably me, but I haven't looked at this particular code in a couple years.  A pasted patch (whether working or not) would make it much clearer.

**Comment by Bulat-Ziganshin** [comment]
Yeah, i looked at this 2-year old thread last night, so my advantage is that i remember all details.

Let's see - non-SIMD approach, omitting unnecessary details, is the cycle:

``` C
int freq[256];
for(p=buf; p<bufend; p++)
    freq[*p]++;
```

where updating freq table translates into asm command

```
   inc [freq+rcx*4]
```

In this command store address can't be computed on Haswell port 7 due to its complex form. Are you remember that? My idea is to replace it with

``` C
char freq[256];
int main_freq[256];
for(p=buf; p<bufend; p++)
    if (++freq[*p] == 0)  
        main_freq[*p] += 256;
for(i=0; i<256; i++)
    main_freq[i] += freq[i];
```

where updating freq table translates into

```
   inc [freq+rcx]
   jz inc_main_freq
```

so rcx*4 in address calculation is replaced with rcx, allowing store address to be computed on port 7. 

The price of my approach is unpredicted jump (i.e. 14+ cycles) to inc_main_freq once per 256 bytes, i.e. 0.054 cycles/byte, which looks even smaller than the price of AVX operations in your fastest code.

**Comment by nkurz** [comment]
OK, I think I understand now.  

Unfortunately, I think you are applying a different definition to "simple address" than Intel.  Their main docs (and Agner) don't define the term at all, but experimentation has shown that it means "single register possibly plus an offset".  The problem is not the scale factor, but the existing of an index register at all.     

I can't point to official Intel docs for this --- it's based on word of mouth and trial and error.  When you search, it's mostly just me and a few other  people posting online (Stephen Canon, Peter Cordes) who detail it.    This might be the closest to anything official you'll find: http://stackoverflow.com/questions/25899395/obtaining-peak-bandwidth-on-haswell-in-the-l1-cache-only-getting-62

Maybe there's some way get the benefit you want by using a constant offset and a fixed mmap()?  But I don't think it's going to be pretty.

**Comment by nkurz** [comment]
Separately, I don't know that that loop is limited by Port 7.  There are two "complex" address generation ports (Ports 3 and 4), and we are only doing one read per cycle.  The only time you really need to force the use of Port 7 is when you are trying to do two read plus a store in the same cycle.  I'm doubtful that allowing things to run on Port 7 here will actually give a speedup in this case.
